### ISSUE TITLE
Add Table of Contents, List of Figures and List of Tables

### ISSUE THREAD

**Issue by yaleLeeNGA** [python-openxml/python-docx#723]
The current version of python-docx does not seem to handle Table of Contents nor List of Figures nor List of Tables at all. So I currently use the trick provided [here](https://stackoverflow.com/questions/51360649/how-to-update-table-of-contents-in-docx-file-with-python-on-linux) to manually create them and update them right after saving the docx file.  
But the trick works on Windows only, with a Word environment.  

Based on my research, many users seem faced to that problem. It would be very usefulif python-docx could handle this cross-platform, application-independant.

**Comment by yaleLeeNGA** [comment]
@ppebay FYI

**Comment by zwggit** [comment]
how to add TOC  on  linux, it is trouble to me !

**Comment by yaleLeeNGA** [comment]
@zwggit If you have Word installed in your Linux environment, and have the possibility to run it in a subprocess, you can use the trick described [here](https://stackoverflow.com/questions/51360649/how-to-update-table-of-contents-in-docx-file-with-python-on-linux).

**Comment by zwggit** [comment]
@yaleLeeNGA 
yes, i see， office it's work,  but my word open by WPS， it's not done.

**Comment by raquelalvz** [comment]
Hello!
How I insert a list of table or list of figures in the same page as table of contents?

**Comment by xiaominzhaoparadigm** [comment]
> Hello!
> How I insert a list of table or list of figures in the same page as table of contents?

Here is the solution
https://github.com/xiaominzhaoparadigm/python-docx-add-list-of-tables-figures/blob/master/LOT.py

**Comment by buhtz** [comment]
For auto-updating fields (toc, numbering in figure/table captions)
https://github.com/elapouya/python-docx-template/issues/151#issuecomment-442722594

**Comment by 1krishnasharma** [comment]
> > Hello!
> > How I insert a list of table or list of figures in the same page as table of contents?
> 
> Here is the solution https://github.com/xiaominzhaoparadigm/python-docx-add-list-of-tables-figures/blob/master/LOT.py

Hi, 
I am new to python-docx, I am using this code but when I open the docs after running code. there is no figure and table in list of table/figure page. Do i need to make figure , table captions as field if yes then how can i do it.
right now I am using this code :- 

def List_of_fig(self,new_doc):
        para = new_doc.add_paragraph("List of Figures")
        para.alignment = WD_ALIGN_PARAGRAPH.CENTER 
        for run in para.runs:
            run.font.name = 'Arial'
            run.font.size = Pt(14)
            run.bold = True
            run.underline = True
        paragraph = new_doc.add_paragraph()
        run = paragraph.add_run()
        fldChar = OxmlElement('w:fldChar')
        fldChar.set(qn('w:fldCharType'), 'begin')
        fldChar.set(qn('w:dirty'), 'true')
        instrText = OxmlElement('w:instrText')
        instrText.set(qn('xml:space'), 'preserve')
        instrText.text = 'TOC \\h \\z \\c "Figure"'  #"Table" of list of table and "Figure" for list of figure
        fldChar2 = OxmlElement('w:fldChar')
        fldChar2.set(qn('w:fldCharType'), 'separate')
        fldChar3 = OxmlElement('w:t')
        fldChar3.text = "Right-click to update field."
        fldChar3 = OxmlElement('w:updateFields') 
        fldChar3.set(qn('w:val'), 'true') 
        fldChar2.append(fldChar3)
        
        fldChar4 = OxmlElement('w:fldChar')
        fldChar4.set(qn('w:fldCharType'), 'end')
        
        run._r.append(fldChar)
        run._r.append(instrText)
        run._r.append(fldChar2)
        run._r.append(fldChar4)

**Comment by 1krishnasharma** [comment]
Hi all,

from the above discussion, I am able to add table of content, list of figure, and list of tables. But the spaces between the fields is little bit large in table of content page and also i want to add heading 1 as bold heading 2 as bold how can i do, it can anyone help me.

@scanny @buhtz @yaleLeeNGA @xiaominzhaoparadigm @raquelalvz  anyone?
